sparta/src/DAG.cpp

```diff
@@ -8,76 +8,98 @@
 
 namespace sparta
 {
-
-    void DAG::CycleException::outPutIssue_(std::ostream & os, bool dot) const
-    {
-        if(dot) {
-            os << "digraph dag_issue {\n";
-            os << "\trankdir=TB;\n";
-            os << "\tnode [shape=record, fontname=Helvetica, fontsize=10];\n\n";
-        }
-        else {
-            os << "DAG CYCLE: " << std::endl;
+    /**
+     * Write a text version of the CycleException's cycle vertex list
+     * @param os
+     */
+    void DAG::CycleException::writeText(std::ostream& os) const {
+        os << "DAG CYCLE: " << std::endl;
+
+        const Vertex* prior_v = nullptr;
+        for (const auto& v : cycle_set_) {
+            if (prior_v != nullptr) {
+                const Edge* e = prior_v->getEdgeTo(v);
+                sparta_assert(e != nullptr);
+                os << " -> " << v->getLabel()
+                   // << "\t// edge: " << e->getLabel()
+                   << std::endl;
+            }
+            os << "\t" << v->getLabel();
+            prior_v = v;
         }
 
-        Vertex * very_first = nullptr;
-        Vertex * very_last = nullptr;
-
-        typename Vertex::VList cycle_set = cycle_set_;
-
-        while(!cycle_set.empty())
-        {
-            Vertex * first = cycle_set.front();
-            cycle_set.pop_front();
-
-            Vertex * second = nullptr;
-            if(!cycle_set.empty()) {
-                second = cycle_set.front();
-            }
-            if(first == second) {
-                continue;
+        // TODO: FOR NOW -- we relax the constraint that the final
+        // vertex in the cycle_set_ needs to have an edge back to
+        // the first vertex.
+        //Vertex* first = cycle_set_.front();
+        //const Edge* e = prior_v->getEdgeTo(first);
+        //sparta_assert(e != nullptr);
+        //os << " -> " << first->getLabel()
+           //<< std::endl;
+
+        // Find and print the cyclic edge from prior_v
+        bool found_cycle_edge = false;
+        for (const auto& w : cycle_set_) {
+            const Edge* e = prior_v->getEdgeTo(w);
+            if (e != nullptr) {
+                os << " -> " << w->getLabel()
+                   << std::endl;
+                found_cycle_edge = true;
+                break;
             }
+        }
+        sparta_assert(found_cycle_edge);
+    }
 
-            if(!very_first) { very_first = first; }
-            if(second) {
-                os << "\t\"" << first->getLabel() << "\"";
-                os << " -> \"" << second->getLabel() << "\"";
-                const Edge * e = first->getEdgeTo(second);
-                if(e != nullptr && dot) {
-                    os << " [fontsize=8 label=\"" << e->getLabel() << "\"]";
-                }
-                os << ";\n";
-                very_last = second;
+    /**
+     * Write a DOT graph version of the CycleException's cycle vertex list
+     * @param os
+     */
+    void DAG::CycleException::writeDOT(std::ostream& os) const {
```

Comment:
Same here -- deeper function name?  Also, can these methods be used to output the DAG generally?  I.e. not necessarily for a DAG issue?

Reply:
No, these are just for the cycle. DAG has separate methods for full DAG output